File: 1582-special-positions-in-a-binary-matrix/1582-special-positions-in-a-binary-matrix.js
/**
 * @param {number[][]} mat
 * @return {number}
 */
var numSpecial = function(mat) {
    let count = 0;
    
    for (let row = 0; row < mat.length; row++){
        for (let column = 0; column < mat[0].length; column++){
            if (mat[row][column] == 1 && isSpecial(mat, row, column)){
                count += 1;
            }
        }
    }
    
    return count;
};


function isSpecial(mat, row, column){
    const rowData = mat[row];
    if (rowData.filter(val => val == 1).length > 1){
        return false;
    }
    
    let upRow = row;
    let downRow = row;

    while (upRow - 1 >= 0) {
        upRow--;
        if (mat[upRow][column] !== 0) {
            return false;
        }
    }

    while (downRow + 1 < mat.length) {
        downRow++;
        if (mat[downRow][column] !== 0) {
            return false;
        }
    }
    
    return true;
}

File: 1582-special-positions-in-a-binary-matrix/1582-special-positions-in-a-binary-matrix.py
class Solution:
    def numSpecial(self, mat: List[List[int]]) -> int:
        count = 0
        
        for r in range(len(mat)):
            for c in range(len(mat[0])):
                if mat[r][c] == 1 and self.isSpecial(mat, r, c):
                    count += 1
        
        return count
    
    def isSpecial(self, mat, row, column):
        rowData = mat[row]
        if rowData.count(1) > 1:
            return False
        
        upRow, downROw = row, row
        while upRow-1 >= 0:
            upRow -= 1
            if mat[upRow][column] != 0:
                return False
        
        while downROw+1 < len(mat):
            downROw += 1
            if mat[downROw][column] != 0:
                return False
        
        return True
        
        